src/php/lib/Grpc/RpcServer.php

```diff
@@ -32,137 +32,120 @@
-            if (!$request_param->getType()) continue;
-            $request_type = $request_param->getType()->getName();
-
-            // $full_path needs to match the incoming event->method
-            // from requestCall() for us to know how to handle the request
-            $this->paths_map[$full_path] = [
-                'service' => $service,
-                'method' => $method_name,
-                'request' => new $request_type(),
-            ];
-        }
+        $this->paths_map = array_merge($this->paths_map, $MethodDescs);
     }
 
-    public function run() {
+    public function run()
+    {
         $this->start();
-        while (true) {
+        while (true) try {
             // This blocks until the server receives a request
             $event = $this->waitForNextEvent();
-            if (!$event) {
-                throw new Exception(
-                    "Unexpected error: server->waitForNextEvent delivers"
-                    . " an empty event");
-            }
-            if (!$event->call) {
-                throw new Exception(
-                    "Unexpected error: server->waitForNextEvent delivers"
-                    . " an event without a call");
-            }
-            $this->call = $event->call;
-            $full_path = $event->method;
-
-            // TODO: Can send a proper UNIMPLEMENTED response in the future
-            if (!array_key_exists($full_path, $this->paths_map)) continue;
 
-            $service = $this->paths_map[$full_path]['service'];
-            $method = $this->paths_map[$full_path]['method'];
-            $request = $this->paths_map[$full_path]['request'];
-
-            $request = $this->loadRequest($request);
-            if (!$request) {
-                throw new Exception("Unexpected error: fail to parse request");
-            }
-            if (!method_exists($service, $method)) {
-                // TODO: Can send a proper UNIMPLEMENTED response in the future
-                throw new Exception("Method not implemented");
+            $full_path = $event->method;
+            $server_writer = new ServerCallWriter($event->call);
+
+            if (!array_key_exists($full_path, $this->paths_map)) {
+                $server_writer->finish(Status::unimplemented());
+                continue;
+            };
+
+            $method_desc = $this->paths_map[$full_path];
+            $context = new ServerContext($event);
+            $server_reader = new ServerCallReader(
+                $event->call,
+                $method_desc->request_type
+            );
+
+            try {
+                $this->processCall(
+                    $method_desc,
+                    $server_reader,
+                    $server_writer,
+                    $context
+                );
+            } catch (\Exception $e) {
+                $server_writer->finish(
+                    Status::status(GRPC_STATUS_INTERNAL, $e->getMessage())
```

Comment:
Oh you already have the `INTERNAL` status here. Oh so the constructor can deal with all the statuses. Then I guess there's no need for those convenience methods for OK and Unimplemented.

Reply:
OK and Unimplemented are used in generated stub